Rewrite `SimpleDidSetRequest` to not require a typechecked body

DougGregor opened the pull request:
SimpleDidSetRequest's implementation was depending on type-checking the body of the observer, meaning that computing the interface type of a `didSet` would require type-checking the body first. This triggers some request cycles, and is unnecessarily complicated.

Rewrite the implementation to work on a non-type-checked body, by using name lookup to establish when the `oldValue` parameter is referenced from the body.

DougGregor committed:
Rewrite `SimpleDidSetRequest` to not require a typechecked body

SimpleDidSetRequest's implementation was depending on type-checking
the body of the observer, meaning that computing the interface type of
a `didSet` would require type-checking the body first. This triggers
some request cycles, and is unnecessarily complicated.

Rewrite the implementation to work on a non-type-checked body, by
using name lookup to establish when the `oldValue` parameter is
referenced from the body.

DougGregor commented:
@swift-ci please smoke test

slavapestov reviewed (approved):
Perhaps one day we can model oldValue as a capture or something, so that interface type computation does not depend on the contents of didSet's body.

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
@swift-ci please clean smoke test macOS

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Don't short-circuit TypeRefinementContext building for functions

`buildTypeRefinementContextHierarchyDelayed()` is already guarded
behind a request, so it doesn't need its own logic for eliminating
redundant computation. Remove that logic, which incorrectly assumed
that the TRC was built for a function if the function body has already
been parsed.

DougGregor commented:
@swift-ci please smoke test
